keps/sig-cluster-lifecycle/kubeadm/1739-customization-with-patches/README.md

```diff
@@ -0,0 +1,626 @@
+breaking changes more visible. Additionally, upgrade instructions will be
+updated adding the recommendation to --dry-run and check expected changes
+before upgrades.
+
+_Patch apply errors_
+
+If a patch fails to apply, kubeadm command should exit with a descriptive
+error.
+
+The users would have to either remove the problematic patch or amend it.
+
+_Confision between the old and new customization features_
+
+For a period of time (at least one release) the flags `--experimental-kustomize`
+and `--experimental-customize` need to co-exist, which can confuse the users.
+
+By marking `--experimental-kustomize` as deprecated, the flag will
+be hidden from command --help` screens. The warning that is printing
+when the deprecated flag is used should be amended to denote that
+`--experimental-customize` should be used instead.
+
+The `--experimental-customize` patches would apply _after_ the
+`--experimental-kustomize` are applied.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Code organization
+
+The code organization of the new feature will be done in a similar
+fashion to the existing Kustomize feature. The backend will be stored
+in a new package called `customize`. Kubeadm commands would have minimal
+exposure to the underlying backend and only call a single method that
+accepts the directory where the user patches are.
+
+### Patch formats
+
+The proposed design is supporting the following patch formats similarly
+to kubectl:
+- Strategic Merge Patches (default):
+  - Supported as `strategic` by `kubectl patch`.
+  - Will be implemented using the `k8s.io/apimachinery` library.
+- RFC6902 JSON patches:
+  - Supported as `json` by `kubectl patch`.
+  - Would be implemented using the `github.com/evanphx/json-patch`
+  library.
+- RFC7396 JSON merge patches:
+  - Supported as `merge` by `kubectl patch`.
+  - Would be implemented using the `github.com/evanphx/json-patch`
+  library.
+
+Note that kubectl already uses the same backends for the above
+patch formats.
+
+Patches defined in both YAML and JSON will be supported, with
+the exception of `json` which must be in JSON only.
+
+Conversion from YAML to JSON will be performed using the
+`sigs.k8s.io/yaml` library.
+
+### Patch file naming format
+
+Once the user passes a folder with patches to a kubeadm command,
+kubeadm must be able to determine what format each file is and what
+manifest it is targeting.
+
+Parsing the patch contents to determine the patch format is not possible,
+because the formats `strategic` and `merge` are the exactly the same,
+while they do have different apply mechanisms.
+
+The proposal is to have the following naming format:
+`componentname[suffix][+patchtype][.extension]`
```

Comment:
My first reaction to this is that we are surfacing a lot of details of patches into the file name.
I would drop the suffix part in favour of support of multiple patches per file, so ultimately the info that are surfacing are only component name and patch type

Reply:
commented here https://github.com/kubernetes/enhancements/pull/1742#discussion_r422716640
but let's discuss further during the kubeadm office hours.